pkg/storage/ingest/writer.go

```diff
@@ -0,0 +1,285 @@
+	logger := log.With(w.logger, "partition", partitionID)
+
+	// Do not export the client ID, because we use it to specify options to the backend.
+	metrics := kprom.NewMetrics("cortex_ingest_storage_writer",
+		kprom.Registerer(prometheus.WrapRegistererWith(prometheus.Labels{"partition": strconv.Itoa(int(partitionID))}, w.registerer)),
+		kprom.FetchAndProduceDetail(kprom.Batches, kprom.Records, kprom.CompressedBytes, kprom.UncompressedBytes))
+
+	return kgo.NewClient(
+		kgo.ClientID(w.kafkaCfg.ClientID),
+		kgo.SeedBrokers(w.kafkaCfg.Address),
+		kgo.RequiredAcks(kgo.AllISRAcks()),
+		kgo.AllowAutoTopicCreation(),
+		kgo.DefaultProduceTopic(w.kafkaCfg.Topic),
+		kgo.DialTimeout(w.kafkaCfg.DialTimeout),
+		kgo.WithHooks(metrics),
+		kgo.WithLogger(newKafkaLogger(logger)),
+
+		// Use a static partitioner because we want to be in control of the partition.
+		kgo.RecordPartitioner(newKafkaStaticPartitioner(int(partitionID))),
+
+		// Set the upper bounds the size of a record batch.
+		kgo.ProducerBatchMaxBytes(16_000_000),
+
+		// A cluster metadata update is a request sent to a broker and getting back the map of partitions and
+		// the leader broker for each partition. The cluster metadata can be updated (a) periodically or
+		// (b) when some events occur (e.g. backoff due to errors).
+		//
+		// MetadataMinAge() sets the minimum time between two cluster metadata updates due to events.
+		// MetadataMaxAge() sets how frequently the periodic update should occur.
+		//
+		// We configure the client to frequently refresh the cluster metadata. This allows us to quickly react in case
+		// some brokers are unhealthy.
+		//
+		// The side effect of frequently refreshing the cluster metadata is that if the backend returns each time a
+		// different authoritative owner for a partition, then each time cluster metadata is updated the Kafka client
+		// will create a new connection for each partition, leading to a high connections churn rate.
+		kgo.MetadataMaxAge(10*time.Second),
+
+		// By default, the Kafka client allows 1 Produce in-flight request per broker. Disabling write idempotency
+		// (which we don't need), we can increase the max number of in-flight Produce requests per broker. A higher
+		// number of in-flight requests, in addition to short buffering ("linger") in client side before firing the
+		// next Produce request allows us to reduce the end-to-end latency.
+		//
+		// The result of the multiplication of producer linger and max in-flight requests should match the maximum
+		// Produce latency expected by the Kafka backend in a steady state. For example, 50ms * 20 requests = 1s,
+		// which means the Kafka client will keep issuing a Produce request every 50ms as far as the Kafka backend
+		// doesn't take longer than 1s to process them (if it takes longer, the client will buffer data and stop
+		// issuing new Produce requests until some previous ones complete).
+		kgo.DisableIdempotentWrite(),
```

Comment:
I think this is true in other clients, but in franz-go, disabling idempotency forces the client to only issue one request at a time -- franz-go favors not duplicating data. _With_ idempotency, franz-go allows 5 requests per broker (technically 4 due to some internal accounting but it's close enough).

Reply:
Oh wait I missed MaxProduceRequestsInflightPerBroker just two lines down, my mistake. Nice!

Reply:
Thanks for looking at it and your feedback!